### ISSUE TITLE
Can not find issuer 'C=US,O=(STAGING) Internet Security Research Group,CN=(STAGING) Doctored Durian Root CA X3' for certificate 'C=US,O=(STAGING) Let's Encrypt,CN=(STAGING) Artificial Apricot R3'

### ISSUE THREAD

**Issue by samcic** [fszlin/certes#248]
We use this library to generate LE certificates for our app (thanks for this library)!

We have a unit test covering the certificate-generating process on the LE staging V2 endpoint.

The code we're using to generate the certificate is:

```
var acmeContext = new AcmeContext(WellKnownServers.LetsEncryptStagingV2, <account key>);
var order = await acmeContext.NewOrder(new[] { hostname });
// (authorization and challenge validation code omitted...working fine)
var privateKey = KeyFactory.NewKey(KeyAlgorithm.ES256);
var cert = await order.Generate(new CsrInfo
                {
                    CountryName = CsrCountryName,
                    State = CsrState,
                    Locality = CsrLocality,
                    Organization = CsrOrganization,
                    OrganizationUnit = CsrOrganizationUnit,
                    CommonName = hostname
                }, privateKey);
var pfxBuilder = cert?.ToPfx(privateKey);
return pfxBuilder?.Build($"letsencrypt-cert-{hostname}", password);
```

Just in the last day or two, we've started seeing our unit test of this code above fail deterministically with the following message:

> "Can not find issuer 'C=US,O=(STAGING) Internet Security Research Group,CN=(STAGING) Doctored Durian Root CA X3' for certificate 'C=US,O=(STAGING) Let's Encrypt,CN=(STAGING) Artificial Apricot R3'."

I'm not the strongest when it comes to cryptography, but based on what I could see, the issuers in [this embedded folder](https://github.com/fszlin/certes/tree/master/src/Certes/Resources/Certificates) are used for the search.

Is it possible that LetsEncrypt made a change and a new one needs to be added now?

**Comment by AndreasHoffmann2** [comment]
Same here.
It is working fine for the real environment, but fails for staging.

**Comment by webprofusion-chrisc** [comment]
Hi, the problem is that to build your PFX chain the code expects to find the root issuers cert in the machine cert store or you need to provide it.

The old issuers were embedded in certes as an assembly resource, which is why it knows about those.

The problem with root certificates and ACME is that acme doesn't declare a way for you to get the root certs when they change, so you have to go get them from the CA and then make them available to whatever uses them.

**Comment by webprofusion-chrisc** [comment]
The code to add a custom issuer is roughly this:

```cs
var issuerDer= <read a issuer pem file from somewhere and convert to binary/der>
var pfx = certificateChain.ToPfx(csrKey);

 pfx.AddIssuers(issuerPem);

var pfxBytes = pfx.Build(certFriendlyName, pwd);

```

The reason the issuer shouldn't really be embedded is there are now many different ACME implementation (public and private) and they all have different root issuer certs. Note that even if you provide the issuer cert the pfx you build won't actually be trusted unless that issuer root is in your clients trust store.

**Comment by samcic** [comment]
@webprofusion-chrisc Nice, thanks for the elaboration!

We grabbed the certificate binary from [here](http://stg-dst3.i.lencr.org/) (as reference from [here](https://community.letsencrypt.org/t/new-issuer-for-letsencrypt-staging/145630/4)), then modified our certificate-generation logic to accept optional additional issuers to add before the final call to `Build` (as per your code). We're just adding this "doctored" issuer in our unit test. Works fine now!

Closing this now.
